torch/_torch_docs.py

```diff
@@ -5885,3 +5885,58 @@ def parse_kwargs(desc):
             [4, 5, 6],
             [4, 5, 6]])
 """)
+
+
+add_docstr(torch.cartesian_prod,
+           r"""
+cartesian_prod(tensors) -> seq
+
+Do cartesian product of the given sequence of tensors. The behavior is similar to
+python's `itertools.prod`. The difference is, if the arguments is a sequence of
+size :math:`k`, `itertools.prod` generate :math:`k`-tuples, while `torch.cartesian_prod`
+create :math:`k` output tensors.
+
+Arguments:
+    tensors (sequence of Tensors): sequence of scalars or 1 dimensional tensors.
+        Scalars will be treated as tensors of size :math:`(1,)` automatically.
+
+Returns:
+    seq (sequence of Tensors): If the input has :math:`k` tensors of size
+        :math:`(N_1,), (N_2,), \ldots , (N_k,)`, then the output would also has :math:`k` tensors,
+        where all tensors are of size :math:`N_1 \times N_2 \times \ldots \times N_k`.
+
+Example::
+
+    >>> a = torch.tensor([1, 2, 3])
+    >>> b = torch.tensor([4, 5])
+    >>> torch.cartesian_prod([a, b])
+    (tensor([1, 1, 2, 2, 3, 3]), tensor([4, 5, 4, 5, 4, 5]))
+""")
+
+
+add_docstr(torch.combinations,
+           r"""
+combinations(tensor, r=2, with_replacement=False) -> seq
+
+Compute combinations of length :math:`r` of the given tensor. The behavior is similar to
+python's `itertools.combinations` when `with_replacement` is set to `False`, and
+`itertools.combinations_with_replacement` when `with_replacement` is set to `True`.
+
+Arguments:
+    tensor (Tensor): the tensor.
+    r (int, optional): number of elements to combine
+    with_replacement (boolean, optional): whether to allow duplication in combination
+
+Returns:
+    seq (sequence of Tensors): :math:`r` tensors.
+
+Example::
+
+    >>> a = torch.tensor([1, 2, 3])
+    >>> torch.combinations(a)
+    (tensor([1, 1, 2]), tensor([2, 3, 3]))
```

Comment:
Should this return `(tensor([1, 2]), tensor(1, 3), tensor(2, 3))`?

Reply:
I think `tensor([[1, 2], [1, 3], [2, 3]])` would be better `(tensor([1, 2]), tensor(1, 3), tensor(2, 3))`, and same argument as above, I'm here returning the unbinded version, but I'm happy to change it to stacked version.

Reply:
What you said is a great point, and I do think we should support this.

My use case is, I have a tensor of shape `(N, A, 3)` that stores the xyz coordinates of `A` atoms of `N` different molecules. For example, the x, y, and z coordinate value of the `j`^th atom of molecule `i` is at `coordinates[i, j, 0]`, `coordinates[i, j, 1]`, and `coordinates[i, j, 2]`.

What I want is to get triplet of atoms, so that I can further compute the angle of these three atoms. I'm doing something like:

```python
indices1, indices2, indices3 = torch.combinations(torch.arange(A), r=3)
# Here I'm assuming the same behavior as the (tensor([1, 1, 2]), tensor([2, 3, 3]))

coordinates1 = coordinates.index_select(1, indices1)
coordinates2 = coordinates.index_select(1, indices2)
coordinates3 = coordinates.index_select(1, indices3)

R12 = coordinates2 - coordinates1
R13 = coordinates3 - coordinates1

cos_angle = torch.nn.functional.cosine_similarity(R12, R13, dim=-1)
angle = torch.acos(cos_angle)
```